pinot-common/src/main/java/org/apache/pinot/common/utils/helix/IdealStateGroupCommit.java

```diff
@@ -124,35 +125,44 @@ public IdealState commit(HelixManager helixManager, String resourceName,
-          IdealState idealState = dataAccessor.getProperty(idealStateKey);
-
-          // Make a copy of the idealState above to pass it to the updater
-          // NOTE: new IdealState(idealState.getRecord()) does not work because it's shallow copy for map fields and
-          // list fields
-          IdealState idealStateCopy = HelixHelper.cloneIdealState(idealState);
-
-          /**
-           * If the local cache does not contain a value, need to check if there is a
-           * value in ZK; use it as initial value if exists
-           */
-          IdealState updatedIdealState = first._updater.apply(idealStateCopy);
-          first._updatedIdealState = updatedIdealState;
-          Iterator<Entry> it = queue._pending.iterator();
-          while (it.hasNext()) {
-            Entry ent = it.next();
-            if (!ent._resourceName.equals(mergedResourceName)) {
-              continue;
+          IdealState response = updateIdealState(helixManager, mergedResourceName, idealState -> {
+            IdealState updatedIdealState = first._updater.apply(idealState);
+            first._updatedIdealState = updatedIdealState;
+            first._exception = null;
+            if (processed.size() > 1) {
+              queue._pending.addAll(processed.subList(1, processed.size()));
+              processed.clear();
+              processed.add(first);
+            }
+            Iterator<Entry> it = queue._pending.iterator();
+            while (it.hasNext()) {
+              Entry ent = it.next();
+              if (!ent._resourceName.equals(mergedResourceName)) {
+                continue;
+              }
+              processed.add(ent);
+              updatedIdealState = ent._updater.apply(updatedIdealState);
+              ent._updatedIdealState = updatedIdealState;
+              ent._exception = null;
+              it.remove();
             }
-            processed.add(ent);
-            updatedIdealState = ent._updater.apply(idealStateCopy);
-            ent._updatedIdealState = updatedIdealState;
-            it.remove();
+            return updatedIdealState;
+          }, retryPolicy, noChangeOk);
+          if (response == null) {
+            RuntimeException ex = new RuntimeException("Failed to update IdealState");
+            for (Entry ent : processed) {
+              ent._exception = ex;
+              ent._updatedIdealState = null;
+            }
+            throw ex;
+          }
+        } catch (Throwable e) {
+          // If the update failed, we should re-add all entries to the queue
+          for (Entry ent : processed) {
+            ent._exception = e;
+            ent._updatedIdealState = null;
           }
-          IdealState finalUpdatedIdealState = updatedIdealState;
-          updateIdealState(helixManager, resourceName, anyIdealState -> finalUpdatedIdealState,
-              retryPolicy, noChangeOk);
+          throw e;
```

Comment:
I think we should only throw if the mergedResourceName matches the one originally requested perhaps?
Otherwise some "bad" znode update would cause totally unrelated failures.

Plus the "owning" thread of the future `Entry` would have returned a failure but the `Entry` might eventually still be processed

Reply:
Previously the exception will directly throw, here we just set exception for all the batch entries. So other threads waiting for these entries won't be blocking.

Reply:
So I think directly throwing was wrong previously too.

The current thread may be initially processing some other resource -- my suggestion is to just remove this `throw e` and let the return/raise at the bottom handle this instead if it is appropriate?

Reply:
I think the resourceName is always the same.

In the catch block, for all the `entries` inside the `processed` ArrayList, they are all belonging to the same resourceName.

The `first` entry is held all the way from the external to internal.

Reply:
hmm so I mean to say --

The caller's interface is this --

```java
public IdealState commit(HelixManager helixManager, String resourceName, Function<IdealState, IdealState> updater, RetryPolicy retryPolicy, boolean noChangeOk);
```

Since we hash out of 100 buckets, there exists some chance (though quite low) that we may need to process some other resource before the Entry we are interested in.

So the caller might be interested in `resourceName <- A` but `mergedResourceName <- B`. So an exceptioni raised while trying to perform `B` should not be thrown to `A` since it is not related to the callers intent.

Reply:
This commit call may process other resource, but because of the while loop here:
![image](https://github.com/user-attachments/assets/ebfc1fcb-a89a-40aa-b3a7-cd5f6f903449)

It will eventually process the entry it pushes, or the entry got processed by other threads already, then this commit thread could just read the result.

Reply:
Right, so imagine your queue slot looks like `[A]`

Some call to `commit(_, B, _)` is made -- `[A, B]`.
Since nothing is running, our thread drives the execution, starting with `A`.

Now, consider if `A` fails and raises an exception.
The queue will look like: `[B]`, but our thread which "owns" the Entry would have already failed.
Worse, some other thread will come and execute our commit.

Reply:
make sense, changed the logic to only operate on its own resource.